### ISSUE TITLE
Mercurial branch test failures

### ISSUE THREAD

**Issue by QuLogic** [airspeed-velocity/asv#755]
I'm trying to build 0.3, but 3 mercurial tests when fail checking branch behaviour. It seems like more revisions are expected than hg produces:
```
_________________________ test_get_branch_commits[hg] __________________________

two_branch_repo_case = (<test.tools.Hg object at 0x7fdc8b979208>, 'default', <asv.plugins.mercurial.Hg object at 0x7fdc8b845080>, <asv.config.Config object at 0x7fdc8b8450f0>)

    def test_get_branch_commits(two_branch_repo_case):
        # Test that get_branch_commits() return an ordered list of commits (last
        # first) and follow first parent in case of merge
        dvcs, master, r, conf = two_branch_repo_case
        expected = {
            master: [
                "Revision 6",
                "Revision 4",
                "Merge stable",
                "Revision 3",
                "Revision 1",
            ],
            "stable": [
                "Revision 5",
                "Merge master",
                "Revision 2",
                "Revision 1",
            ],
        }
        for branch in conf.branches:
            commits = [
                dvcs.get_commit_message(commit_hash)
                for commit_hash in r.get_branch_commits(branch)
            ]
>           assert commits == expected[branch]
E           AssertionError: assert ['Revision 6'... 'Revision 1'] == ['Revision 6',... 'Revision 1']
E             At index 4 diff: 'Revision 2' != 'Revision 1'
E             Left contains more items, first extra item: 'Revision 1'
E             Use -v to get the full diff

/builddir/build/BUILD/asv-0.3/test/test_repo.py:284: AssertionError
_____________________ test_get_new_branch_commits[hg-all] ______________________

two_branch_repo_case = (<test.tools.Hg object at 0x7fdc8b967d68>, 'default', <asv.plugins.mercurial.Hg object at 0x7fdc8b967550>, <asv.config.Config object at 0x7fdc8b967128>)
existing = []
expected = ['Revision 6', 'Revision 4', 'Merge stable', 'Revision 3', 'Revision 1', 'Revision 5', ...]

    @pytest.mark.parametrize("existing, expected", [
        # No existing commit, we expect all commits in commit order,
        # master branch first
        ([], ["Revision 6", "Revision 4", "Merge stable", "Revision 3",
              "Revision 1", "Revision 5", "Merge master", "Revision 2"]),
    
        # New commits on each branch
        (["Revision 4", "Merge master"], ["Revision 6", "Revision 5"]),
    
        # No new commits
        (["Revision 6", "Revision 5"], []),
    
        # Missing all commits on one branch (case of new branch added in config)
        (["Revision 6"], ["Revision 5", "Merge master", "Revision 2", "Revision 1"]),
    ], ids=["all", "new", "no-new", "new-branch-added-in-config"])
    def test_get_new_branch_commits(two_branch_repo_case, existing, expected):
        dvcs, master, r, conf = two_branch_repo_case
    
        existing_commits = set()
        for branch in conf.branches:
            for commit in r.get_branch_commits(branch):
                message = dvcs.get_commit_message(commit)
                if message in existing:
                    existing_commits.add(commit)
    
        assert len(existing_commits) == len(existing)
    
        new_commits = r.get_new_branch_commits(conf.branches, existing_commits)
        commits = [dvcs.get_commit_message(commit) for commit in new_commits]
>       assert commits == expected
E       AssertionError: assert ['Revision 6'...ision 1', ...] == ['Revision 6',...ision 5', ...]
E         At index 4 diff: 'Revision 2' != 'Revision 1'
E         Use -v to get the full diff

/builddir/build/BUILD/asv-0.3/test/test_repo.py:316: AssertionError
__________ test_get_new_branch_commits[hg-new-branch-added-in-config] __________

two_branch_repo_case = (<test.tools.Hg object at 0x7fdc8b858f60>, 'default', <asv.plugins.mercurial.Hg object at 0x7fdc8b858710>, <asv.config.Config object at 0x7fdc8b858860>)
existing = ['Revision 6']
expected = ['Revision 5', 'Merge master', 'Revision 2', 'Revision 1']

    @pytest.mark.parametrize("existing, expected", [
        # No existing commit, we expect all commits in commit order,
        # master branch first
        ([], ["Revision 6", "Revision 4", "Merge stable", "Revision 3",
              "Revision 1", "Revision 5", "Merge master", "Revision 2"]),
    
        # New commits on each branch
        (["Revision 4", "Merge master"], ["Revision 6", "Revision 5"]),
    
        # No new commits
        (["Revision 6", "Revision 5"], []),
    
        # Missing all commits on one branch (case of new branch added in config)
        (["Revision 6"], ["Revision 5", "Merge master", "Revision 2", "Revision 1"]),
    ], ids=["all", "new", "no-new", "new-branch-added-in-config"])
    def test_get_new_branch_commits(two_branch_repo_case, existing, expected):
        dvcs, master, r, conf = two_branch_repo_case
    
        existing_commits = set()
        for branch in conf.branches:
            for commit in r.get_branch_commits(branch):
                message = dvcs.get_commit_message(commit)
                if message in existing:
                    existing_commits.add(commit)
    
        assert len(existing_commits) == len(existing)
    
        new_commits = r.get_new_branch_commits(conf.branches, existing_commits)
        commits = [dvcs.get_commit_message(commit) for commit in new_commits]
>       assert commits == expected
E       AssertionError: assert ['Revision 5'...ision 2', ...] == ['Revision 5',... 'Revision 1']
E         At index 2 diff: 'Revision 4' != 'Revision 2'
E         Left contains more items, first extra item: 'Revision 3'
E         Use -v to get the full diff
```
Using `mercurial-4.7-1.fc30.x86_64` and `python3-hglib-2.6.1-3.fc29.noarch` in Rawhide.

**Comment by pv** [comment]
It passes on earlier mercurial/hglib versions, so probably something changed there.

**Comment by QuLogic** [comment]
I tested with Fedora 27 (`mercurial-4.2.3-1.fc27.x86_64` and `python3-hglib-2.6.1-1.fc27.noarch`) which fails other things, but not these tests.

I tested with Fedora 28 (`mercurial-4.4.2-4.fc28.x86_64` and `python3-hglib-2.6.1-1.fc28.noarch`) which works.

I tested with Fedora 29 (`mercurial-4.5.3-1.fc29.x86_64` and `python3-hglib-2.6.1-3.fc29.noarch`) which fails.

**Comment by pv** [comment]
Apparently the issue is that in `hg log` the `--follow-first` option (deprecated?) apparently doesn't do anything any more. Mercurial 4.5 [changelog](https://www.mercurial-scm.org/wiki/WhatsNew) mentions some changes but the feature seems broken now...

**Comment by pv** [comment]
fix in #756

**Comment by QuLogic** [comment]
Thanks! This [works](https://koji.fedoraproject.org/koji/taskinfo?taskID=30351792).
